subprojects/core/src/main/java/org/gradle/internal/work/DefaultAsyncWorkTracker.java

```diff
@@ -60,16 +65,23 @@ public void waitForCompletion(Operation operation) {
         }
 
         try {
-            for (AsyncWorkCompletion item : workItems) {
-                try {
-                    item.waitForCompletion();
-                } catch (Throwable t) {
-                    failures.add(t);
-                }
-            }
+            if (workItems.size() > 0) {
+                projectLockService.withoutProjectLock(new Runnable() {
+                    @Override
+                    public void run() {
+                        for (AsyncWorkCompletion item : workItems) {
```

Comment:
I've changed this so that we can query an `AsyncWorkCompletion` to see if it's complete or not, so we check first and then only release project locks if we find that something is still in progress.